tasks/indexing/task_ipni.go

```diff
@@ -0,0 +1,452 @@
+			if err != nil {
+				return false, xerrors.Errorf("failed to get private libp2p key for miner %d: %w", p.SpID, err)
+			}
+
+			pkey, err := crypto.UnmarshalPrivateKey(privKey)
+			if err != nil {
+				return false, xerrors.Errorf("unmarshaling private key: %w", err)
+			}
+
+			pubK := pkey.GetPublic()
+			pid, err := peer.IDFromPublicKey(pubK)
+			if err != nil {
+				return false, fmt.Errorf("getting peer ID: %w", err)
+			}
+
+			pcid, err := cid.Parse(p.PieceCid)
+			if err != nil {
+				return false, xerrors.Errorf("parsing piece CID: %w", err)
+			}
+
+			pi := abi.PieceInfo{
+				PieceCID: pcid,
+				Size:     p.Size,
+			}
+
+			b := new(bytes.Buffer)
+			err = pi.MarshalCBOR(b)
+			if err != nil {
+				return false, xerrors.Errorf("marshaling piece info: %w", err)
+			}
+
+			_, err = tx.Exec(`SELECT insert_ipni_task($1, $2, $3, $4, $5, $6, $7, $8)`, p.SpID,
+				p.Sector, p.Proof, p.Offset, b.Bytes(), false, pid.String(), id)
+			if err != nil {
+				if harmonydb.IsErrUniqueContraint(err) {
+					ilog.Infof("Another IPNI announce task already present for piece %s in deal %s", p.PieceCid, p.UUID)
+					// SET "complete" status to true for this deal, so it is not considered next time
+					n, err := tx.Exec(`UPDATE market_mk12_deal_pipeline SET complete = TRUE WHERE uuid = $1`, p.UUID)
+					if err != nil {
+						return false, xerrors.Errorf("store IPNI success: updating pipeline: %w", err)
+					}
+					if n != 1 {
+						return false, xerrors.Errorf("store IPNI success: updated %d rows", n)
+					}
+					// We should commit this transaction to set the value
+					stop = false // we found a task to schedule, keep going
+					return true, nil
+				}
+				if strings.Contains(err.Error(), "already published") {
+					ilog.Infof("Piece %s in deal %s is already published", p.PieceCid, p.UUID)
+					// SET "complete" status to true for this deal, so it is not considered next time
+					n, err := tx.Exec(`UPDATE market_mk12_deal_pipeline SET complete = TRUE WHERE uuid = $1`, p.UUID)
+					if err != nil {
+						return false, xerrors.Errorf("store IPNI success: updating pipeline: %w", err)
+					}
+					if n != 1 {
+						return false, xerrors.Errorf("store IPNI success: updated %d rows", n)
+					}
+					// We should commit this transaction to set the value
+					stop = false // we found a task to schedule, keep going
+					return true, nil
+				}
+				return false, xerrors.Errorf("updating IPNI announcing task id: %w", err)
+			}
+
+			stop = false // we found a task to schedule, keep going
+			return true, nil
+		})
+	}
+
+	return nil
+}
+
+func (I *IPNITask) Adder(taskFunc harmonytask.AddTaskFunc) {}
+
+func (I *IPNITask) GetSpid(db *harmonydb.DB, taskID int64) string {
+	var spid string
+	err := db.QueryRow(context.Background(), `SELECT sp_id FROM ipni_task WHERE task_id = $1`, taskID).Scan(&spid)
+	if err != nil {
+		log.Errorf("getting spid: %s", err)
+		return ""
```

Comment:
Is it OK to return an empty value, or should an error be returned?

Reply:
Yes. That one is used by UI for showing which task is running on which node. Empty will simply results in one task missing from the list instead of an error for the whole page.